### PR TITLE
feat(publishing): Enable Giscus widgets in published notes

### PR DESCRIPTION
**Pull request by hikchoi**
# feat(publishing): Enable Giscus widgets in published notes

This PR: 
- Adds Giscus related configuration under `publishing.giscus` config namespace
- Adds Giscus widget component
- Adds a per-note frontmatter flag `enableGiscus` used to enable Giscus widgets in respective notes

# Pull Request Checklist
### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended
- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics
- [~] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended
- [~] can we track the performance of this change to know if it is _successful_?
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

### Basics

- [~] [Write Tests](dev.process.qa.test) 
- [x] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [~] Common cases tested
- [~] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended
- [x] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](dev.ref.test-workspace)
- CSS
  - [x] display is correct for following dimensions
    - [x] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [x] lg: screen ≥ 992px
    - [x] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [x] display is correct for following browsers (across the various dimensions)
    - [x] safari
    - [x] firefox
    - [x] chrome

## Docs
- [ ] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
- [~] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

### Extended
- [ ]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg: [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

### CODE REVIEW

**Comment by hikchoi** [comment]
NOTE:

1. this adds `@giscus/react` to `nextjs-template`
2. Configs are not added to the UI yet for timely delivery

File: packages/common-all/src/types/configs/publishing/giscus.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
+// https://github.com/giscus/giscus-component/blob/main/react/src/lib/types.ts
```
Comment (hikchoi) [inline packages/common-all/src/types/configs/publishing/giscus.ts:1]: How should I go about properly attributing the project here?

They don't have a separate type package that I could use and it's a bit weird to add `@giscus/react` to `common-all` just for the types.

File: packages/common-all/src/types/configs/publishing/giscus.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+// non-trivial / const related type definitions are breaking JSON schema generation
```
Comment (hikchoi) [inline packages/common-all/src/types/configs/publishing/giscus.ts:3]: There are some generic typing on Giscus' end that our JSON schema generator fails.
Leaving this to be fixed on the next phase.

**Comment by Harshita-mindfire** [comment]
> NOTE:
> 2. Configs are not added to the UI yet for timely delivery
@hikchoi : heads up the docs to update config are here: [[Add New Config|dendron://dendron.docs/pkg.common-all.dev.cook.add-new-config]]. Also, if we are opting it to update in next iteration, feel free to create a TODO task for the same.

**Review by jonathanyeung — APPROVED** [review]
Tested locally, works! Apart from Harshita's comment above, looks good!

**Comment by hikchoi** [comment]
> > NOTE:
> > 2. Configs are not added to the UI yet for timely delivery
> 
> @hikchoi : heads up the docs to update config are here: [[Add New Config|dendron://dendron.docs/pkg.common-all.dev.cook.add-new-config]]. Also, if we are opting it to update in next iteration, feel free to create a TODO task for the same.

Thanks for the heads up. I just am not sure if we want to add this to the UI at all yet, and this is quite an urgent feature so I'm deferring it to later once it's discussed fully.

I'll create a separate task for it 👍🏼

**Pull request merged**